### PR TITLE
Annotations improvements and bugfixes

### PR DESCRIPTION
**Pull request by damianhxy**
## Description
**Bug fixes**
- Prevent selection of autograded problems when editing annotation
- Persist shared comments checkbox for global annotations
- Preserve newlines in descriptions
- Add handling for `nil` problem `max_score`

**New features**
- Display current max score on edit problem page
- Add problem max score to annotation panel
- Allow multiple annotations per line

**Breaking changes**
- Make shared comments operate on a per-problem basis

**Others**
- Optimize performance of updating grades and annotation panes

## Motivation and Context
Addresses the following feedback from 122:

1. Keep newlines in the global annotations when displayed. In the screenshot below, I input A, B, C, D on separately lines, but to students, it’ll be all on one line, which is undesirable
2. Make comments added to the shared pool specific to a problem (or not, should be user defined). I currently see that if a comment is added to one problem, it’ll show up in the shared pool for other problems too. This might introduce a lot of clutter if we want problem-specific shared comments.
3. Add the max number of points for each problem to the annotations panel
4. When I hover a line that already has a comment, the green + button appears, making me assume that I can add another comment, but clicking the button doesn’t do anything. It’d be nice to have multiple comments per line. This is depicted in the screenshot above as well.

**Note**
Feedback included the following, to be addressed in a future PR:

Point values should stay with a shared comment added to the pool. In the screenshot below, "A1: Has contract”’s point values when I created it, +1.0, should be automatically applied when I choose this comment from the shared pool. This issue comes up not only when I attach it to another Problem, but within the same Problem as well.

## How Has This Been Tested?

**1. Prevent selection of autograded problems when editing annotation**
- On an autograded assessment, make an annotation (ensure that we can't select the autograded problem)
- [**previously**] Edit the annotation, we are able to select the autograded problem
- [**now**] Edit the annotation, we are **not** able to select the autograded problem

**2. Persist shared comments checkbox for global annotations**
- Make a global annotation, mark it as a shared comment
- [**previously**] Edit the annotation, the shared comment checkbox is not checked
- [**now**] Edit the annotation, the shared comment checkbox **is** checked

**3. Preserve newlines in descriptions**
- Create a multi-line global annotation, check that newlines are preserved in the annotation pane
- Create a multi-line normal annotation, check that newlines are preserved in the annotation pane
- Also observe that annotations descriptions start to the immediate right of the point badge as usual

Screenshot 2022-12-02 at 00 38 18

**4. Display current max score on edit problem page**
- [**previously**] Edit an existing problem, observe that "max score" field shows 0
- [**now**] Edit an existing problem, observe that "max score" field shows the current max value
- Create a new problem, ensure the page renders, and the "max score" field shows 0

**5. Add problem max score to annotation panel**
- Observe that annotation pane displays problem score and max score next to current delta. i.e. `<delta> (<current score> / <max score>)`
- Update a problem's max score, ensure the number changes accordingly
- Create a new problem, ensure the number displays accordingly
- Test with positive grading too

Negative Grading
Screenshot 2022-12-02 at 00 39 00

Positive Grading
Screenshot 2022-12-02 at 00 39 37

**6. Allow multiple annotations per line**
- If user clicks add button multiple times without first cancelling / submitting the first open form, observe that we do not create any more forms and display a warning message
Screenshot 2022-12-01 at 20 16 42

- Successfully create multiple annotations per line
- Ensure that "edit" forms do not count towards this limit. That is, we can be editing multiple annotations on the same line at once, but even so, we can still click the add button to start creating a new annotation.
- Check that the page automatically scrolls to the new annotation form

**7. Make shared comments operate on a per-problem basis**
- Create a bunch of normal / global shared annotations
- Start creating a new annotation, ensure that the correct shared annotations show up for the autocomplete (i.e. only the shared annotations from the same problem). Ensure that the autocomplete options change when the problem dropdown is changed.
- Edit an existing annotation, ensure that the autocomplete options change accordingly too.

**8. Optimize performance of updating grades and annotation panes**
- Make an annotation, ensure that grades and annotation panes update correctly

**9. Add handling for `nil` problem `max_score`**
- Remove `:max_score` from the `validates :name, :max_score, presence: true` line in `models/problem.rb`
- Create a problem with no max score (disable the required validation on the field)
- [**previously**] An error is thrown (from trying to `printf` `nil`)
- [**now**] View a submission, page should load with a `-` in the place of max score on the annotations and grades pane

Screenshot 2022-12-02 at 12 19 47

Screenshot 2022-12-02 at 12 19 50

- Also check gradesheet and submission history table

Screenshot 2022-12-02 at 15 31 57

Screenshot 2022-12-02 at 15 32 05

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [x] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [x] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

**Comment by 20wildmanj** [comment]
probably existing but the cursor doesn't change when hovering over / clicking the plus button for adding a new annotation , so it's hard to know if I created a new annotation, esp when I'm adding to the same line and there are already multiple annotations

**Comment by 20wildmanj** [comment]
Clicking "update annotation" on an existing annotation when another annotation is also open removes the other open annotation
Screen Shot 2022-12-02 at 12 27 22

**Comment by 20wildmanj** [comment]
when autofilling, you can add the same comment that got autofilled back to the shared comment pool, and it'll become a new shared comment

**Comment by 20wildmanj** [comment]
adding newlines to an annotation makes the shared comments drop-down get misaligned
Screen Shot 2022-12-02 at 12 35 32

File: docs/features/annotations.md [hunk 1]
```diff
@@ -12,9 +12,9 @@ Annotations is a feature introduced as part of the Speedgrader update to Autolab
 Hover over any line of the code and click on the green arrow, and the annotation form will appear. Add the comment, adjust the score, and select the targeted problem.
 
 When the "Add to shared comment pool" option is checked, the comment will be saved to a shared comment pool.
-When creating new annotations, shared comments matching what you have typed will be available via a dropdown list.
+When creating new annotations, shared comments matching what you have typed will be available via a dropdown list. Only the 50 most recently added shared comments will be displayed.
```
Comment (20wildmanj) [inline docs/features/annotations.md:15]: haha good to specify how many shared comments will be displayed, 50 is quite a lot though so hopefully most people will never need to remember

**Review by 20wildmanj — APPROVED** [review]
Specified bug-fixes all work well, and discussed offline that other issues will be resolved at a later date as they are pre-existing and are not easy fixes. LGTM

**Comment by damianhxy** [comment]
> probably existing but the cursor doesn't change when hovering over / clicking the plus button for adding a new annotation , so it's hard to know if I created a new annotation, esp when I'm adding to the same line and there are already multiple annotations

Added code to scroll to the new annotation form when a user clicks on the plus button.

**Comment by damianhxy** [comment]
### Summary of remaining issues
> Clicking "update annotation" on an existing annotation when another annotation is also open removes the other open annotation

Known issue, it's because of how `annotations.js` liberally calls `displayAnnotations()` after some events (change file, cancel, update), which basically rerenders all the annotation boxes and thus closes the existing ones.

We need a more targeted approach of only editing the required annotation.

> when autofilling, you can add the same comment that got autofilled back to the shared comment pool, and it'll become a new shared comment

Given the current design of shared comments (`shared_comment` flag on annotations), there isn't really a way around this atm.

An alternative design might be to have a table of shared annotations where each annotation points to the shared annotation it is utilizing and perhaps the shared annotation can store a refcount. This could be a little complicated though.

> Multi-line comments don't seem to open all the way when you re-open / autofill

Known issue, some previous observations:
Screenshot 2022-12-02 at 15 52 44


> adding newlines to an annotation makes the shared comments drop-down get misaligned

Good observation, the textarea is pretty buggy in general.

### Other notes

Implementing the remaining feedback as mentioned in the PR description is a little more involved. It would necessitate the writing of custom code (see https://jqueryui.com/autocomplete/#custom-data) to attach extra attributes to each dropdown element (i.e. the score) and to fill in the score fields when we select a value.

A problem doesn't get a score until we make an annotation for it. We should see if it's possible to automatically initialize a score for each problem for new submissions, to reduce the risk of calling undefined methods on `nil`.

When we click on an annotation on the right side bar, we start scrolling from the top of the code viewer each time, even if the annotation is already in view. This can result in janky animation.

CC: @20wildmanj

**Comment by 20wildmanj** [comment]
Scroll works, other tweaks also work, LGTM

**Pull request merged**